SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp

```diff
@@ -106,37 +109,50 @@ Worker_ComponentData USpatialSender::CreateLevelComponentData(AActor* Actor)
 	return ComponentFactory::CreateEmptyComponentData(SpatialConstants::NOT_STREAMED_COMPONENT_ID);
 }
 
-void USpatialSender::SendAddComponent(USpatialActorChannel* Channel, UObject* Subobject, const FClassInfo& SubobjectInfo, uint32& OutBytesWritten)
+void USpatialSender::SendAddComponentForSubobject(USpatialActorChannel* Channel, UObject* Subobject, const FClassInfo& SubobjectInfo, uint32& OutBytesWritten)
 {
 	FRepChangeState SubobjectRepChanges = Channel->CreateInitialRepChangeState(Subobject);
 	FHandoverChangeState SubobjectHandoverChanges = Channel->CreateInitialHandoverChangeState(SubobjectInfo);
 
 	ComponentFactory DataFactory(false, NetDriver, USpatialLatencyTracer::GetTracer(Subobject));
 
 	TArray<FWorkerComponentData> SubobjectDatas = DataFactory.CreateComponentDatas(Subobject, SubobjectInfo, SubobjectRepChanges, SubobjectHandoverChanges, OutBytesWritten);
+	SendAddComponents(Channel->GetEntityId(), SubobjectDatas);
+
+	Channel->PendingDynamicSubobjects.Remove(TWeakObjectPtr<UObject>(Subobject));
+}
 
-	for (int i = 0; i < SubobjectDatas.Num(); i++)
+void USpatialSender::SendAddComponents(Worker_EntityId EntityId, TArray<FWorkerComponentData> ComponentDatas)
+{
+	if (ComponentDatas.Num() == 0)
 	{
-		FWorkerComponentData& ComponentData = SubobjectDatas[i];
-		Connection->SendAddComponent(Channel->GetEntityId(), &ComponentData);
+		return;
 	}
 
-	Channel->PendingDynamicSubobjects.Remove(TWeakObjectPtr<UObject>(Subobject));
+	// Update ComponentPresence.
+	check(StaticComponentView->HasAuthority(EntityId, SpatialConstants::COMPONENT_PRESENCE_COMPONENT_ID));
+	ComponentPresence* Presence = StaticComponentView->GetComponentData<ComponentPresence>(EntityId);
+	Presence->AddComponentDataIds(ComponentDatas);
+	FWorkerComponentUpdate Update = Presence->CreateComponentPresenceUpdate();
+	Connection->SendComponentUpdate(EntityId, &Update);
+
+	for (FWorkerComponentData& ComponentData : ComponentDatas)
+	{
+		Connection->SendAddComponent(EntityId, &ComponentData);
+	}
 }
 
 void USpatialSender::GainAuthorityThenAddComponent(USpatialActorChannel* Channel, UObject* Object, const FClassInfo* Info)
 {
-	const FClassInfo& ActorInfo = ClassInfoManager->GetOrCreateClassInfoByClass(Channel->Actor->GetClass());
-	const WorkerAttributeSet WorkerAttribute{ ActorInfo.WorkerType.ToString() };
-	const WorkerRequirementSet AuthoritativeWorkerRequirementSet = { WorkerAttribute };
-
-	EntityAcl* EntityACL = StaticComponentView->GetComponentData<EntityAcl>(Channel->GetEntityId());
+	Worker_EntityId EntityId = Channel->GetEntityId();
 
 	TSharedRef<FPendingSubobjectAttachment> PendingSubobjectAttachment = MakeShared<FPendingSubobjectAttachment>();
 	PendingSubobjectAttachment->Subobject = Object;
 	PendingSubobjectAttachment->Channel = Channel;
 	PendingSubobjectAttachment->Info = Info;
 
+	// We collect component IDs related to the dynamic subobject being added to gain authority over.
+	TArray<Worker_ComponentId> NewComponentIds;
```

Comment:
Can we reserve the size upfront? (size of Info->SchemaComponents[Type] ?)

Reply:
Should just be `NewComponentIds.SetNum(SCHEMA_Count);`?

Reply:
Alas no

Reply:
I've got nothing